include/boost/url/detail/impl/remove_dot_segments.ipp

```diff
@@ -0,0 +1,518 @@
+        // - E leaves "/" or happens at the end
+        std::size_t out = 0;
+        std::size_t n = 0;
+        while (!s.empty())
+        {
+            n = detail::pct_decode_starts_with_unchecked(s, "../");
+            if (n)
+            {
+                out += 3;
+                s.remove_prefix(n);
+                continue;
+            }
+            n = detail::pct_decode_starts_with_unchecked(s, "./");
+            if (n)
+            {
+                s.remove_prefix(n);
+                continue;
+            }
+            break;
+        }
+
+        // D.  if the input buffer consists only
+        // of "." or "..", then remove that from
+        // the input buffer; otherwise, ...
+        // Rule D can only happen after A is consumed:
+        // - B and C write "/" to the output
+        // - D can only happen at the end
+        // - E leaves "/" or happens at the end
+        if (detail::pct_decode_compare_unchecked(s, ".") == 0)
+        {
+            s = {};
+        }
+        else if (detail::pct_decode_compare_unchecked(s, "..") == 0)
+        {
+            out += 2;
+            s = {};
+        }
+        return out;
+    };
+    string_view s0 = s0_init;
+    string_view s1 = s1_init;
+    std::size_t s0_prefix_n = remove_prefix(s0);
+    std::size_t s1_prefix_n = remove_prefix(s1);
+
+    auto pop_last = [](
+            string_view& s,
+            string_view& c,
+            std::size_t& l,
+            bool r)
+    {
+        c = {};
+        std::size_t n = 0;
+        while (!s.empty())
+        {
+            // B.  if the input buffer begins with a
+            // prefix of "/./" or "/.", where "." is
+            // a complete path segment, then replace
+            // that prefix with "/" in the input
+            // buffer; otherwise,
+            n = detail::pct_decode_ends_with_unchecked(s, "/./");
+            if (n)
+            {
+                c = s.substr(s.size() - n);
+                s.remove_suffix(n);
+                continue;
+            }
+            n = detail::pct_decode_ends_with_unchecked(s, "/.");
+            if (n)
+            {
+                c = s.substr(s.size() - n);
+                s.remove_suffix(n);
+                continue;
+            }
+
+            // C. if the input buffer begins with a
+            // prefix of "/../" or "/..", where ".."
+            // is a complete path segment, then
+            // replace that prefix with "/" in the
+            // input buffer and remove the last
+            // segment and its preceding "/"
+            // (if any) from the output buffer
+            // otherwise,
+            n = detail::pct_decode_ends_with_unchecked(s, "/../");
+            if (n)
+            {
+                c = s.substr(s.size() - n);
+                s.remove_suffix(n);
+                ++l;
+                continue;
+            }
+            n = detail::pct_decode_ends_with_unchecked(s, "/..");
```

Comment:
ah is this like `string_view::ends_with` but with applying percent-decoding?

Reply:
Exactly.